CODING_GUIDELINES.md

```diff
@@ -0,0 +1,101 @@
+
+Sphinx supports [reStructuredText][sphinx-rest] (reST) markup language to add additional formatting options to the generated docs, however section [_3.8 Comments and Docstrings_](https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings) in the Google Python Style Guide does not define how to use markups in docstrings. Therefore, we decided to forbid the use of reST markup in docstrings except for:
+
+   - Cross-referencing other objects using Sphinx text roles for the [Python domain](https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#the-python-domain) (as explained [here](https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#python-roles)).   
+   - Very basic formatting markup to improve _readability_ of the generated documentation without obscuring the source docstring (e.g. ``` ``literal`` ```  strings).
+   
+Regarding code examples in docstrings, we highly encourage to use [doctest][doctest] format to automatically test they are in sync with the code.
+
+### Module structure
+
+In general, you should structure new Python modules in the following way:
+
+1. (Only for **executable scripts**) _shebang_ line: `#! /usr/bin/env python3`.
+2. License header boilerplate (check `LICENSE_HEADER.txt`).
+3. Module docstring.
+4. Imports, alphabetically ordered within each block:
+   1. Block of imports from the standard library.
+   2. Block of imports from general third party libraries using standard shortcuts when customary (e.g. `numpy as np`).
+   3. Block of imports from specific modules of the project.
+5. (Optional, mainly for re-exporting symbols from other modules) Definition of exported symbols:   
+```python
+__all__ = ["func_a", "CONST_B"]
+   ```
+6. Public constant and typing definitions.
+7. Module contents organized in a meaningful way for reading and understanding the module, usually defining functions before classes.
+
+Try to keep sections and items within sections ordered logically, adding comments to make it explicit if needed (also in configuration files). If there is not one single evident logical order, just pick the order you consider best or use alphabetical order.
+
+### Ignoring QA errors
+
+You may ocassionally need to disable checks from _quality assurance_  (QA) tools (e.g. linters, type checkers, etc.) on specific lines because the tool is not able to fully understand why that piece of code is needed. This is usually feasible inlining a special comment like `# type: ignore`. You should **only** ignore QA errors when you fully understand its cause and it is not reasonable to fix it by rewriting the offending code in a different way. Additionally, you should add a brief comment to make sure anyone else reading the code also understands what is happening there. For example:
```

Comment:
```suggestion
You may ocassionally need to disable checks from _quality assurance_  (QA) tools (e.g. linters, type checkers, etc.) on specific lines because the tool is not able to fully understand why that piece of code is needed. This is usually done by a special comment like `# type: ignore`. You should **only** ignore QA errors when you fully understand its cause and rewriting it to pass QA checks would actually make it less readable. Additionally, you should add a brief comment to make sure anyone else reading the code also understands what is happening there. For example:
```